### PR TITLE
[JENKINS-62806] compatibility fix for Jenkins UI changes

### PR DESCRIPTION
**Pull request by imavroukakis**
Fixes issues with configuration UI not rendering in Jobs encountered in Jenkins 2.264+

### CODE REVIEW

**Comment by imavroukakis** [comment]
I'm aware it's mega-hacky, happy to follow any recommendations, influenced heavily by how the dropbox jenkins plugin fixed the same issue ;-)

**Comment by kinow** [comment]
I was aware of the issue, but thought it would be an epic change, so decided to postpone until I had a good amount of spare time to learn how to fix it. You probably saved me already a ton of time, and we might be able to ship it sooner to users thanks to your PR @imavroukakis .

I will take a look over the next days, but not too worried if it's hacky or not, as long as we can improve it over time without causing issues to users.

Thanks!

**Review by kinow — COMMENTED** [review]
Testing first on `master` to get a "broken" environment. Set the version to `2.269` as the greatest LTS release is `2.263.1` (I think `2.264+` is required for divs?).

The `mvn clean install` command fails due to tests that are not able to locate elements in the form. Which is good, at least we have tests that confirm whether the plug-in basic behaviour is working after a change.

Started it again, now with `mvn clean install -DskipTests`, followed by `mvn hpi:run`.

The first job I tried (an existing job) worked fine.

Screenshot from 2020-12-06 18-46-28

Unfortunately I couldn't create a new test job, as I had the following error when saving it.

```
Caused by: java.lang.IllegalArgumentException: Failed to instantiate class hudson.model.ParameterDefinition from {"randomName":"","name":"city","script":{"value":"0","script":{"script":"return ['sao paulo', 'rio', 'auckland']","sandbox":false},"fallbackScript":{"script":"return ['city error']","sandbox":false},"stapler-class":"org.biouno.unochoice.model.GroovyScript","$class":"org.biouno.unochoice.model.GroovyScript"},"stapler-class":["org.biouno.unochoice.model.ScriptlerScript","org.biouno.unochoice.ChoiceParameter"],"$class":["org.biouno.unochoice.model.ScriptlerScript","org.biouno.unochoice.ChoiceParameter"],"description":"","choiceType":"PT_SINGLE_SELECT","filterable":false,"filterLength":"1"}
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:681)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:728)
	at org.kohsuke.stapler.RequestImpl.bindJSON(RequestImpl.java:478)
	at org.kohsuke.stapler.RequestImpl.instantiate(RequestImpl.java:787)
	... 87 more
Caused by: java.lang.IllegalArgumentException: Class ["org.biouno.unochoice.model.ScriptlerScript","org.biouno.unochoice.ChoiceParameter"] is specified in JSON, but no such class found in classLoader hudson.PluginManager$UberClassLoader
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:674)
	... 90 more
Caused by: java.lang.ClassNotFoundException: cached miss for ["org.biouno.unochoice.model.ScriptlerScript","org.biouno.unochoice.ChoiceParameter"]
	at hudson.PluginManager$UberClassLoader.findClass(PluginManager.java:2028)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at org.kohsuke.stapler.RequestImpl$TypePair.convertJSON(RequestImpl.java:669)
	... 90 more
```

I guess it might be caused by https://www.jenkins.io/blog/2020/11/10/spring-xstream/ (found in [another issue on JIRA](https://issues.jenkins.io/browse/JENKINS-64189)). Might need more time to continue testing later :+1:

**Comment by imavroukakis** [comment]
Interesting, I didn't see this behaviour when I patched the plug-in on our CI instance, will also do a bit of digging. Thanks!

**Comment by kinow** [comment]
Thanks @imavroukakis . I didn't have a lot of time, so I could have missed something - hopefully nothing too obvious. But if you could try it too. I want to have the broken environment, so that I can confirm:

1) your branch fixes it and
2) your change also works OK with older versions

Once that's done, it should be ready to be merged and released asap to avoid issues when LTS version is increased.

**Comment by imavroukakis** [comment]
> Thanks @imavroukakis . I didn't have a lot of time, so I could have missed something - hopefully nothing too obvious. But if you could try it too. I want to have the broken environment, so that I can confirm:
> 
> 1. your branch fixes it and
> 2. your change also works OK with older versions
> 
> Once that's done, it should be ready to be merged and released asap to avoid issues when LTS version is increased.

You're welcome! I think I have a solution to the issue, seems like some additional work is needed around the use of `<td>`. I've managed to make one test pass, working on the rest

**Comment by imavroukakis** [comment]
With the changes in [`1a17e1c` (#40)](https://github.com/jenkinsci/active-choices-plugin/pull/40/commits/1a17e1c7f63405f09a1ce5c50fd952d64f736732) tests pass for both Jenkins version `2.204.1` and `2.264`. I think it might be useful to supply `jenkins.version` as a command line parameter and create an additional build, so that both versions can be tested

**Comment by kinow** [comment]
Looks promising @imavroukakis ! WIll try it again next time I have some spare time to review/test it. Hopefully a new release can be cut soon :crossed_fingers: thanks!

File: src/test/java/org/biouno/unochoice/jenkins_cert_1954/TestParametersXssVulnerabilities.java [hunk 1]
```diff
@@ -38,19 +39,20 @@
...
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertNotNull;
 
 /**
  * Tests against XSS. See SECURITY-1954, and SECURITY-2008.
  * @since 2.5
  */
 public class TestParametersXssVulnerabilities {
 
+    private static final VersionNumber TRANSITION_TO_DIV_VERSION = new VersionNumber("2.264");
```
Comment (kinow) [inline src/test/java/org/biouno/unochoice/jenkins_cert_1954/TestParametersXssVulnerabilities.java:55]: :+1: nice!

File: src/main/resources/org/biouno/unochoice/common/checkboxContent.jelly [hunk 2]
```diff
@@ -1,10 +1,54 @@
 <?jelly escape-by-default='true' ?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define"
   xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form"
-  xmlns:i="jelly:fmt" xmlns:p="/lib/hudson/project">
+  xmlns:i="jelly:fmt" xmlns:p="/lib/hudson/project" xmlns:transitionWrapper="transitionWrapper">
+    <d:taglib uri="transitionWrapper">
+      <d:tag name="randomTableIdBlock">
+          <j:choose>
+              <j:when test="${divBasedFormLayout}">
```
Comment (kinow) [inline src/main/resources/org/biouno/unochoice/common/checkboxContent.jelly:8]: Was wondering where that `divBasedFormLayout` was being defined. Answer: Jenkins sets in the page context if the version uses divs, and not tables.

https://github.com/jenkinsci/jenkins/blob/bdb5dce566566e8724679564fe612130c064580f/core/src/main/java/hudson/Functions.java#L305

**Comment by kinow** [comment]
Code looks excellent! Nice job @imavroukakis ! Just need to test it locally. If there's nothing wrong, we should be ready to merge it.

**Comment by kinow** [comment]
Here's how I tested. I used a job I created recently to work on some other issue, with three parameters (1 reacting to the other two).

Then first used the `pom.xml` as-is, and got the following (see table elements).

Screenshot from 2020-12-20 16-19-22

Then applied this simple diff

```diff
diff --git a/pom.xml b/pom.xml
index 91b096f..4d9c5e5 100644
--- a/pom.xml
+++ b/pom.xml
@@ -36,7 +36,7 @@
     <properties>
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <phantomjs.executable>phantomjs</phantomjs.executable>
-        <jenkins.version>2.204.1</jenkins.version>
+        <jenkins.version>2.269</jenkins.version>
         <java.level>8</java.level>
     </properties>

```

And tested the same job (note, no table elements, only div's).

Screenshot from 2020-12-20 16-28-20

Worked like a charm! Clever workaround @imavroukakis , and I found other plug-ins doing something similar. So +1, LGTM and allow users to update Jenkins without breaking their jobs. Release coming up on these holidays, just triaging other issues first to see if I can include a bit more.

Thanks a lot!
Bruno

**Review by kinow — APPROVED** [review]
@imoutsatsos this fixes the style issues when using the latest version of Jenkins. Now Active Choices should work with any version. Thanks to @imavroukakis.

Cheers
Bruno

**Comment by kinow** [comment]
Merged. Changelog added to README too. JIRA updated. Thanks again @imavroukakis !

**Comment by imavroukakis** [comment]
> Merged. Changelog added to README too. JIRA updated. Thanks again @imavroukakis !

My pleasure 😄

**Comment by timja** [comment]
Could this be released please? there's another report in Jira at https://issues.jenkins.io/browse/JENKINS-64568?focusedCommentId=403206&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-403206

**Comment by kinow** [comment]
Hi @timja, I was going to release a new version last week, but got a couple issues blocking the release. I'm having a meeting next week to discuss it with @imoutsatsos, to try to work out these blockers :+1:

**Comment by timja** [comment]
Sure, documentation is here: https://www.jenkins.io/doc/developer/publishing/releasing/

most issues are caused by an old password, as they were all reset nearly a year ago.

**Comment by saurabh-agarwals** [comment]
> Hi @timja, I was going to release a new version last week, but got a couple issues blocking the release. I'm having a meeting next week to discuss it with @imoutsatsos, to try to work out these blockers 👍

any update on this @kinow

**Comment by kinow** [comment]
> > Hi @timja, I was going to release a new version last week, but got a couple issues blocking the release. I'm having a meeting next week to discuss it with @imoutsatsos, to try to work out these blockers +1
> 
> any update on this @kinow

I think most of the work for the next release is done @saurabh-agarwals my bad for taking this long to cut a new release and publish to Jenkins update center. I just need to finish testing it, and waiting for a review/test of one more issue. Then once I find spare time will do the upload to the update center.

Will update this issue once it's released :+1: (but if I take too long feel free to bump the thread)

**Comment by kinow** [comment]
Hi all :wave: 

2.5.2 includes the table-to-div fix contributed by @imavroukakis. The latest version is **2.5.3 which includes a security bug fix** I was working with the Jenkins security team (CVE-2021-21616). So I recommend updating to this latest version if possible.

As always, remember to 

- back up
- use a testbed or dev server
- test :+1: 

Thanks, and sorry for the delay in releasing this change.

Bruno

**Pull request closed without merging**